File: src/trl/sig.ts

export interface ReadSig<T> {
    get (): T
}

export interface WriteSig<T> {
    set (val: T): Und
    setWith (fnUpdate: (val: T) => T): Und
}

export class Sig<T> implements ReadSig<T>, WriteSig<T> {
    private _val: T

    constructor (val: T) {
        this._val = val
    }

    static new<T> (val: T): Sig<T> {
        return new Sig<T>(val)
    }

    get (): T {
        return this._val
    }

    set (val: T): Und {
        this._val = val
    }

    setWith (fnUpdate: (val: T) => T): Und {
        this._val = fnUpdate(this.get())
    }
}

File: src/trl/sig.rs
use ::core::ops::Fn;

pub trait ReadSig<T> {
    fn get() -> T;
}

pub trait WriteSig<T> {
    fn set(val: T);
    fn set_width<F>(fn_update: F)
    where
        F: Fn(T) -> T;
}

pub struct Sig<T> {
    val: T,
}

impl<T> Sig<T> {
    fn new(val: T) -> Self {
        Self { val }
    }
}

impl<T> ReadSig<T> for Sig<T> {
    fn get(&self) -> T {
        self.val
    }
}

impl<T> WriteSig<T> for Sig<T> {
    fn set(&self, val: T) {
        self.val = val
    }

    fn set_width<F>(&self, fn_update: F)
    where
        F: Fn(T) -> T,
    {
        self.val = fn_update(self.val)
    }
}
